src/printer.js

```diff
@@ -2022,6 +1978,76 @@ function printMemberChain(node, options, print) {
   }
 }
 
+function printJSXElement(path, options, print) {
+  const n = path.getValue();
+  var openingLines = path.call(print, "openingElement");
+
+  let elem;
+  if (n.openingElement.selfClosing) {
+    assert.ok(!n.closingElement);
+
+    elem = openingLines;
+  } else {
+    var children = [];
+
+    path.map(
+      function(childPath) {
+        var child = childPath.getValue();
+
+        if (
+          namedTypes.Literal.check(child) && typeof child.value === "string"
+        ) {
+          if (/\S/.test(child.value)) {
+            const beginBreak = child.value.match(/^\s*\n/);
+            const endBreak = child.value.match(/\n\s*$/);
+
+            children.push(
+              beginBreak ? hardline : "",
+              child.value.replace(/^\s+|\s+$/g, endBreak ? "" : " "),
+              endBreak ? hardline : ""
+            );
+          } else if (/\n/.test(child.value)) {
+            children.push(hardline);
+          }
+        } else {
+          children.push(print(childPath));
+        }
+      },
+      "children"
+    );
+
+    var mostChildren = children.slice(0, -1);
+    var closingLines = path.call(print, "closingElement");
+
+    const firstChild = children[0];
+    const lastChild = util.getLast(children);
+    const beginBreak = firstChild && firstChild.type === "line";
+    const endBreak = lastChild && lastChild.type === "line";
```

Comment:
That's definitely helpful to know about, though in this case I _think_ the lines could be inserted by `prettier`, not just from the original source. So I think it wouldn't be adequate in this case. 

But it would be very helpful when deciding whether `{' '}` needs to be inserted, if that's a route we go.